### ISSUE TITLE
Automatic PR reviews

### ISSUE THREAD

**Issue by iangcarroll** [HackathonHackers/personal-sites#207]
We should have a simple bot that uses the GitHub Webhook and API features to mark pull requests as ready to merge or to automatically point out issues. Travis is a good reference on what I want to do.

Ideally, it would:
- Make sure the URL returns a 200/300 response code
- Make sure the list is in ABC order
- Query VirusTotal to make sure it's not malicious
- anything else?

**Comment by Jamonek** [comment]
Lets create it!

**Comment by bryangarza** [comment]
Any update on this?

**Comment by iangcarroll** [comment]
Yeah - I'm in DC for the week so I'll finish it up next week.

**Comment by zmughal** [comment]
++

**Comment by zmughal** [comment]
Also check for duplicates and make all the GitHub URLs normalised to use `https`.

**Comment by robbyoconnor** [comment]
I support this

**Comment by zmughal** [comment]
Also check if someone does a pull request, suggest that their GitHub profile should be in the GitHub list too. That would save me some time since I often go through and add all the PR profiles myself.

**Comment by iangcarroll** [comment]
Flow seems to be:
webhook for pull request => get commit hash from webhook => push pending status to github => run checks on the commit's diff => push good/bad status to github

Gonna have to work out how to parse the diff.

**Comment by iangcarroll** [comment]
https://github.com/iangcarroll/pr-bot/pull/9 shows a demonstration of statuses. Webhook is partially working now.

**Comment by zmughal** [comment]
Do you need to parse the diff? I would apply the patch using `git-apply` or `patch` and then see if the file checker fails on the changed file.

**Comment by zmughal** [comment]
Now that I think about it, you could write something really quickly that just runs on Travis-CI.

**Comment by iangcarroll** [comment]
That's a good idea, but we don't have a file checker (yet).

**Comment by iangcarroll** [comment]
Given the webhook logic is mostly done, moving it to Travis would be more work. Applying the `.patch` version would be interesting, though.

**Comment by iangcarroll** [comment]
URL extraction and testing via `preg_match_all` is done, although cURL currently throws an exception if the domain doesn't exist that we're not catching (and thus the PR gets stuck on pending).

What's left:
- VT result check
- ABC order

**Comment by iangcarroll** [comment]
VT results done: https://github.com/iangcarroll/pr-bot/pull/21

**Comment by iangcarroll** [comment]
Until I wrap this up, I put the source up at https://github.com/iangcarroll/pr-bot-src, if anyone wants to submit a PR for ABC ordering.

**Comment by robbyoconnor** [comment]
PHP?

**Comment by iangcarroll** [comment]
It's a Laravel 5.0 app, yes.

**Comment by robbyoconnor** [comment]
:( Node.js is the only real dev language -- everything else...is just lame :P Esp. php.

**Comment by tjhorner** [comment]
Do we really need something as heavy as a MVC app? We can just have a worker in the background.

**Comment by tjhorner** [comment]
Written in Node.js.

**Comment by iangcarroll** [comment]
Heavy, maybe, but its maintainability, speed or functionality are not impeded by it. All the logic is in `app/Http/Controllers/PullRequestController.php`, for those inclined to contribute.

I'm running into a problem parsing the `.diff`. What I think I'm going to do is perform alphabetical sorting on the sections in it, and not the entire file. This will validate sorting on the three names above and below each addition, which I believe is acceptable. The reason for this is that the .diff is not a complete version of the file, and only stores additions, deletions, and the 3 lines above and below any aforementioned changes. This also dodges the problem of changes to both the personal site and GitHub section.

The only other way would be fetching the resulting Markdown from the commit and parsing that, which seems a bit much (given we'd still have to parse the `.diff` to extract the URL additions so we don't re-check every URL every PR). 

I'll have to re-do part of the validation logic, but I should have something shortly.

**Comment by hellyeah** [comment]
What will it take to bring this to fruition?

**Comment by iangcarroll** [comment]
I'm way too busy (as usual, unfortunately) but I'll lay out the approach I took for future takers per @hellyeah's request:
* Receive a GitHub webhook, verify its signature and discard anything that isn't a newly opened pull request 
* Pull the commit hash from the webhook and fetch it via the GitHub API
* Download the `.diff` from the previous response
* (flawed approach that I took) Regex the .diff for URLs and run checks on them
(correct approach) Parse each line, assuming the last "word" is the URL and everything else is the name, and check the alphabetical order on the names

Of course, both approaches suffer from the 3-line limitation on how diffs are created. @zmughal had suggested applying the .diff to a local file which could work, but was not especially easy with PHP.

I'll remove myself as the assigned person.
